### PR TITLE
split config management into secrets and settings

### LINKED ISSUE #14: Use env for managing secrets
The config system is a convenient way to manage settings that differ for each environment; however, it isn't a secure way to manage secrets. Either we move away from the config infra completely or integrate env into the existing system.

If we want to integrate env, we could create a new `Env` type to our config infra that reads the environment variable name from the `configs/<config>.json` entry (e.g. `$GOOGLE_API_KEY`) and loads the actual value from the environment.

### PR DESCRIPTION
**Pull request by leifwalsh**
fixes #14

TODO:
- [x] Audit deploy script, see if there are things we need to change there (probably at least handle setting `AUTOMATION_ENV`)
- [x] Write usage docs, probably in the readme
- [x] Make sure we have all the right settings committed, and API keys uploaded to Secret Manager
- [x] Test deploying each function
- [x] Test local development
- [x] Fix CI tests to not try to fetch secrets during test discovery

### CODE REVIEW

File: automation/clients/airtable.py [hunk 1]
```diff
@@ -132,6 +133,25 @@ def validate_status(cls, v):
         return v
 
 
+############
```
Comment (leifwalsh) [inline automation/clients/airtable.py:137]: I'm not sure whether I prefer the secrets and settings definitions near where they're used, or centralized in one module? What do you all think?

File: automation/clients/auth0.py [hunk 2]
```diff
@@ -18,19 +17,42 @@ def is_unauthorized(exception):
...
-        self, conf: Auth0Config, secrets_client: secrets.SecretsClient
+        self, secrets_client=SecretsClient(), settings=Auth0Settings()
     ):
-        self._base_url = "https://" + conf.domain
+        self._base_url = "https://" + settings.domain
         self._api_url = self._base_url + "/api/v2"
-        self._client_id = conf.client_id
-        self._client_secret = conf.client_secret
+        self._client_id = settings.client_id
+        self._client_secret = settings.client_secret
```
Comment (leifwalsh) [inline automation/clients/auth0.py]: Should `client_secret` actually be in the secrets?

File: automation/clients/auth0.py [hunk 3]
```diff
@@ -18,19 +17,42 @@ def is_unauthorized(exception):
...
         self._secrets_client = secrets_client
-        self._token = None
+        self._secret = None
+
+    @property
```
Comment (leifwalsh) [inline automation/clients/auth0.py:60]: I haven't tested this property logic yet, but I'm pretty sure it's sound.

File: automation/clients/sendgrid.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+from ..secrets import BaseSecret, SecretsClient
+
+
+class SendgridSecrets(BaseSecret):
+
+    _secret_name = "sendgrid"
+    api_key: SecretStr
+
+
+# Named like a class so it looks like a constructor, but this doesn't manage
```
Comment (leifwalsh) [inline automation/clients/sendgrid.py:13]: It was a) sort of weird we didn't have this in one place yet, and b) also weird that each function that sends email uses one single "sendgrid settings" in the config that determines from and reply-to. Now, the sendgrid client doesn't know what you want from or reply-to to be, each function should have its own settings for that.

File: automation/clients/slack.py [hunk 5]
```diff
@@ -77,7 +78,22 @@ def get_handle(self):
...
+class SlackSecrets(BaseSecret):
+
+    _secret_name = "slack"
+    api_key: pydantic.SecretStr
+    scim_api_key: pydantic.SecretStr
+    resend_invite_secret: pydantic.SecretStr
+
+
+class SlackSettings(pydantic.BaseSettings):
+    test_user_email: Optional[str]
```
Comment (leifwalsh) [inline automation/clients/slack.py:91]: It was sort of weird to me that we have these test fields in the config. The slack test is disabled and seems like it's only run manually anyway. Maybe we can get rid of these eventually?

File: automation/clients/slack.py [hunk 6]
```diff
@@ -87,7 +103,7 @@ class SlackErrors(enum.Enum):
 def generate_password(length):
     charset = string.ascii_letters + string.digits
     return "".join(
-        [random.SystemRandom().choice(charset) for _ in xrange(length)]
+        [random.SystemRandom().choice(charset) for _ in range(length)]
```
Comment (leifwalsh) [inline automation/clients/slack.py:106]: drive-by python 2 removal

File: automation/functions/members.py [hunk 7]
```diff
@@ -6,7 +6,7 @@
 log = structlog.get_logger("poll_members")
 
 
-def on_new(member, *, slack_client, sendgrid_client, auth0_client, from_email):
+def on_new(member, *, slack_client, sendgrid_client, auth0_client):
```
Comment (leifwalsh) [inline automation/functions/members.py:9]: This was just unused

File: automation/tests/helpers.py [hunk 8]
```diff
@@ -2,44 +2,22 @@
...
-from .. import config
-from ..models import MemberModel
 from ..clients import slack
+from ..models import MemberModel
+
+
+TEST_ENV = "environments/test.env"
+
+
+class MockSecretsClient:
```
Comment (leifwalsh) [inline automation/tests/helpers.py:12]: Not the most elegant way to do it, but we needed something so the airtable client constructor wouldn't explode.

File: automation/tests/test_slack.py [hunk 9]
```diff
@@ -1,15 +1,24 @@
...
 
 import pytest
 
+from .helpers import TEST_ENV, MockSecretsClient
+from ..clients import slack
+
+
+TEST_SECRETS_CLIENT = MockSecretsClient(
+    slack=json.dumps(
+        {"api_key": "", "scim_api_key": "", "resend_invite_secret": ""}
```
Comment (leifwalsh) [inline automation/tests/test_slack.py:11]: In order to run this test, we'd need actual values here. I'm not sure what we should do with this test, I'm sure it was useful during development but maybe we can delete it at this point?

File: automation/config.py [hunk 10]
```diff
@@ -1,66 +0,0 @@
-import json
```
Comment (leifwalsh) [inline automation/config.py:1]: Moved to settings, but also each separate client's config moved into the appropriate client module.

File: automation/secrets.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    @classmethod
+    def load(cls, secrets_client=SecretsClient()):
+        """Load a secret from Secret Manager."""
+        data = secrets_client.get_secret(cls._secret_name)
+        secrets = cls.parse_raw(data)
+        secrets._secrets_client = secrets_client
+        return secrets
+
+    @staticmethod
+    def _plaintext_encode(value):
```
Comment (leifwalsh) [inline automation/secrets.py:80]: It's possible to set some `Config` options that will make `.json()` serialize all secret values in plaintext, but I'm doing it this way so that only in `save()` do we serialize plaintext values. This way, if someone accidentally logs `secret.json()` somewhere, the secret values will be sanitized in logs.

File: automation/secrets.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        self._secrets_client.set_secret(self._secret_name, payload)
+
+
+class SendgridSecrets(BaseSecret):
+
+    _secret_name = "sendgrid"
+    api_key: SecretStr
+
+
+if __name__ == "__main__":
```
Comment (leifwalsh) [inline automation/secrets.py:105]: Just for debugging, doesn't actually print values, just validates that you can load them.

File: automation/settings.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+                env_prefix = "foo_"
+
+    This way, in .env files or in the process environment, `foo_my_var` will
+    control that setting.
+    """
+
+    env_file = f"environments/{AUTOMATION_ENV}.env"
+
+
+class GoogleCloudSettings(pydantic.BaseSettings):
```
Comment (leifwalsh) [inline automation/settings.py:32]: I didn't have a better place to put this, and it's needed to bootstrap the secrets manager API calls. It's also not a bad working example when looking at this module.

File: automation/tables.py [hunk 14]
```diff
@@ -1,27 +1,52 @@
...
+from .secrets import SecretsClient
 
 
 class PollableTable(abc.ABC):
 
     table_spec = ...
 
-    def __init__(self, conf: config.Config, read_only=False):
-        self.conf = conf
+    def __init__(
```
Comment (leifwalsh) [inline automation/tables.py:14]: This changed a bunch. Now, there's not a single master config object that gets piped down everywhere:

* All secrets should be loaded from the default `SecretsClient`, generally, developers shouldn't be handling secret material anywhere themselves. If you really want to provide your own `SecretsClient` (like the airtable tests do), you can.
* All settings load their default values from the .env file (chosen by what's in the process's `AUTOMATION_ENV` variable, which defaults to `prod` for now), but if you really want to override them, at any level you can pass in your own settings object. But in normal usage, all this config is loaded implicitly by the `BaseSettings` default constructor somewhere, if you do nothing. So now, `main.py` doesn't need to manage config at all.

Some principles:

* If a function needs a client, it should expect one to be passed in, either by something in this module (one of the pollers, getting it from a mixin), or by one of the local dev scripts constructing one.
* If a client or a function needs settings, it should accept a settings parameter, but should have a default value which uses the no-arg constructor, which means "load it from the environment". This way, callers don't need to provide one, but may override settings if they want, and during development you can override individual settings with environment variables.
* Each layer should generally only accept constructed clients, and also settings for itself, as parameters: one layer shouldn't accept settings for another layer and forward them along, it should just be passed an already configured client for the lower layer.

File: environments/prod.env [hunk 15]
```diff
@@ -0,0 +1,21 @@
+airtable_base_id=appoJ59KQ73mhoPpu
+airtable_table_names='{
+    "inbound": "Phone/Text Inbound (For call back!)",
+    "members": "Volunteers - Form Responses",
+    "volunteer": "Intake Volunteer Sign Up Sheet",
+    "intake": "Intake Tickets",
+    "items_by_household_size": "Items By Household Size"
+}'
+
+auth0_domain=
```
Comment (leifwalsh) [inline environments/prod.env]: I didn't have all the values to put here, actually, we need to review them and make sure everything's in before we roll these out. 

We also need to make sure we have all our API keys in Secret Manager, and audit that they are all appropriately scoped, and replicated across the prod and test projects appropriately (and we probably want separate API keys for each, at least for airtable).

**Comment by leifwalsh** [comment]
I tested that I could run automation.functions.delivery's `__main__` test thing using `dev.env`, and it did what I expected.

File: automation/clients/slack.py (hunk 5 above)
Comment (maybejustmaybe) [reply automation/clients/slack.py:91]: Down to get rid of it; pretty sure its leftover from my initial implementation. But I would like to have a good story for slack integration tests. Maybe let's make an issue for it and punt for now (i.e. delete or deprecate the tests)?

File: automation/secrets.py (hunk 11 above)
Comment (maybejustmaybe) [reply automation/secrets.py:80]: Very helpful explanation! Could you copy-paste it as an inline comment?

File: automation/tests/test_slack.py (hunk 9 above)
Comment (maybejustmaybe) [reply automation/tests/test_slack.py:11]: (see comment above)

File: automation/clients/airtable.py (hunk 1 above)
Comment (maybejustmaybe) [reply automation/clients/airtable.py:137]: I think centralizing in one module will make the settings easier to manage. The downside is that devs may have to modify the settings or secrets modules to write new functionality? Imo some high level documentation and references to existing usages should keep the dev experience pretty straightforward.

Comment (leifwalsh) [reply automation/clients/airtable.py:137]: Yeah, the upside of split config is when you make changes to a function and its expected settings, it's localized to one module (plus you have to update the env files). The upside of centralized config is you can see the expected config in one module. I was leaning toward split config since we still have examples of the whole config in the env files (come to think of it we should have a test that the env files are all valid), but don't feel strongly.

File: automation/clients/slack.py (hunk 5 above)
Comment (leifwalsh) [reply automation/clients/slack.py:91]: #76, let's delete this test later, maybe when we address it.

**Comment by leifwalsh** [comment]
I've tested the intake poller in staging and have deployed it to prod (so emails will come from deliveries@). I believe in it.

File: automation/clients/auth0.py (hunk 2 above)
Comment (leifwalsh) [reply automation/clients/auth0.py]: @cubeghost?

File: environments/prod.env (hunk 15 above)
Comment (leifwalsh) [reply environments/prod.env]: @cubeghost @maybejustmaybe

File: automation/clients/auth0.py (hunk 2 above)
Comment (cubeghost) [reply automation/clients/auth0.py]: yeah definitely should be, it's an API key. raises an interesting question with `client_id` too- it's not necessarily a secret but it's paired with `client_secret` for the authentication flow, so it kind of feels more appropriate to put in secrets
